Rewrite generate_lm.py to allow usage with other languages.

DanBmh opened the pull request.

lissyx commented:
Thanks, this is a discussion we already had and the outcome was that this would add too much complexity.

lissyx commented:
I guess we'll have to wait until @reuben is back

DanBmh commented:
> Thanks, this is a discussion we already had and the outcome was that this would add too much complexity.

I don't think i added much complexity. Only the possibility to use your own txt.file with sentences. And i added some command line arguments for customization.

lissyx commented:
> > Thanks, this is a discussion we already had and the outcome was that this would add too much complexity.
> 
> I don't think i added much complexity. Only the possibility to use your own txt.file with sentences. And i added some command line arguments for customization.

I'm not judging, I'm reporting status of discussions on the exact same feature since I did it for french.

lissyx commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,135 @@
-               '--prune', '0', '0', '1'
+        args.kenlm_bins + 'lmplz', '--order', '5',
+        '--temp_prefix', args.output_dir,
+        '--memory', '75%',
```
maybe you would want that as a parameter

lissyx commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,135 @@
-               '--text', data_lower,
-               '--arpa', lm_path,
-               '--prune', '0', '0', '1'
+        args.kenlm_bins + 'lmplz', '--order', '5',
```
Same, order might require to be a parameter

lissyx commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,135 @@
+    # Convert and count words
+    print('\nConverting to lowercase and counting word occurrences ...')
+    with io.TextIOWrapper(io.BufferedWriter(gzip.open(data_lower, 'w+')), encoding='utf-8') as lower:
+        with open(args.input_txt, encoding='utf8') as upper:
```
This is weird, you open `lower.txt.gz`, decode it and then write that to `args.input_txt` which is documented to be the source text. I think it's not what you intended.

DanBmh replied:
?
```
with io.TextIOWrapper(io.BufferedWriter(gzip.open(data_lower, 'w+')), encoding='utf-8') as lower:
        with open(args.input_txt, encoding='utf8') as upper:
            for line in tqdm.tqdm(upper):
                line_lower = line.lower()
                counter.update(line_lower.split())
                lower.write(line_lower)
```

I'm opening `args.input_txt` as file `upper` and writing every line into `lower.txt.gz`

But I think I will rename those files for better readability when I'm solving the other suggestions you made.

lissyx replied:
Ah, right, well this is confusing, as you can see. I don't understand why you write to `.txt.gz`

DanBmh replied:
This is because i am calling `lmplz ... --text data_lower`. It was called with `.txt.gz` before so I assumed i also need a `.txt.gz` file.

lissyx commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,135 @@
-def main():
-  # Grab corpus.
-  url = 'http://www.openslr.org/resources/11/librispeech-lm-norm.txt.gz'
+import tqdm
```
We already have a dependency against `progressbar2`, can we avoid multiplying the dependencies for the same feature?

DanBmh commented:
My new commits should solve your requested changes.

I also added some statistics about the input file:

```
Calculating word statistics ...
  Your text file has 803288729 words in total
  It has 973673 unique words
  Your top-500000 words are 99.9354 percent of all words
  Your most common word "the" occurred 49059384 times
  The least common word in your top-k is "corders" with 2 times
  The first word with 3 occurrences is "zungwan" at place 420186

```

DanBmh commented:
The repository from @tilmankamp (i think you ment [this](https://github.com/mozilla/oscarlm)) seems to be specifically for the OSCAR corpora. 

What about splitting the file into two scripts? One to download the english text corpus and the other to generate a language model out of a text file with sentences. This would be language/dataset independent then.

reuben commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,147 @@
+        type=str,
+        required=True
+    )
+    parser.add_argument(
+        '--top_k',
+        help='Use top_k most frequent words for the vocab.txt file',
+        type=int,
+        default=500000
+    )
+    parser.add_argument(
+        '--kenlm_bins',
+        help='File path to the kenlm binaries lmplz, filter and build_binary',
+        type=str,
+        default='/DeepSpeech/native_client/kenlm/build/bin/'
+    )
+    parser.add_argument(
+        '--arpa_order',
+        help='Order of k-grams in arpa-file generation',
+        type=int,
+        default=5
+    )
+    parser.add_argument(
+        '--max_arpa_memory',
+        help='Maximum allowed memory usage in arpa-file generation',
+        type=str,
+        default='75%'
+    )
```
Every default here is arbitrary. Make them optional and document the exact command line used to generate the model in `data/lm/kenlm.scorer`.

DanBmh replied:
I tried to use the values you used to generate the model. The path to the kenlm bins is the installation path in the docker container.

Not sure what you mean with optional, if you don't add the parameter to your command, the default values are used, else it will be overwritten with the value from the command.

reuben replied:
Sorry, I meant to say make them required.

DanBmh replied:
Without default values? So that the user has to fill in a value for each single parameter?

reuben replied:
Correct, and then document the invocation that has the values used for the model in the repo.

DanBmh replied:
Ok, i can do this.
Why do you think that is better than using default values? I would prefer default values because i think its more readable, add only flags for params i want to change or which are needed to run this program. Like its also handled in `DeepSpeech.py`.

reuben replied:
Because other users might have different needs? This is the goal of this PR, right? To make the script more flexible.

DanBmh replied:
Yes, but with both of our ways you get a flexible script. I only see a difference in readability and usability. For example in my version i only need to add `--top_k 300000` if i am not happy with `--top_k 500000`. In your version i always have to add  `--top_k 500000`.
Personally i prefer the shorter commands, but its only a design issue. If you think the users should list all parameters, i will move the default values to the example call for the model used here.

reuben commented on data/lm/generate_package.py:
```diff
@@ -112,23 +112,27 @@ def main():
         "--force_utf8",
         default="",
-        help="Boolean flag, force set or unset UTF-8 mode in the scorer package. If not set, infers from the vocabulary.",
+        help="Boolean flag, force set or unset UTF-8 mode in the scorer package. If not set, infers from the vocabulary. Using this wrong can result in a 'Segmentation fault' error at the model evaluation.",
```
Using the wrong scorer can lead to inconsistent behavior but not segmentation fault. If you've seen that happen, please create an issue with steps to reproduce.

DanBmh replied:
Created an issue: https://github.com/mozilla/DeepSpeech/issues/2875

reuben commented on data/lm/generate_package.py:
```diff
@@ -112,23 +112,27 @@ def main():
         "--default_beta",
         type=float,
-        required=True,
+        default=1.85,
```
Don't embed defaults here (aren't you trying to make these scripts more flexible?)

DanBmh commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,147 @@
-    # Build pruned LM.
-    lm_path = os.path.join(tmp, 'lm.arpa')
-    print('Creating ARPA file...')
+    data_lower = os.path.join(args.output_dir, 'lower.txt')
```
I removed it exactly for the case the script fails midway. So that you can look at the files to search for the error.

reuben replied:
Ah, OK.

reuben commented on data/lm/README.rst:
```diff
@@ -1,32 +1,23 @@
 .. code-block:: bash
 
     wget http://www.openslr.org/resources/11/librispeech-lm-norm.txt.gz -O librispeech.txt.gz
     gunzip librispeech.txt.gz
```
nit: Remove.

DanBmh replied:
Delete the intro text again?

reuben replied:
No no, just the `gunzip` part since the script can now read from the .gz version.

reuben commented on data/lm/README.rst:
```diff
@@ -1,32 +1,23 @@
 
 .. code-block:: bash
 
     python3 data/lm/generate_lm.py --input_txt path/to/vocab_sentences.txt --output_dir path/lm/
```
This command should fully specify all the parameters matching the included LM.

reuben replied:
And since `generate_lm.py` now takes a parameter with the path to KenLM tools instead of relying on `$PATH`, this README should at least link to the KenLM build documentation and say that those steps need to be followed first.

DanBmh replied:
I still don't like the use of the parameters without default values. I understand we should specify all the parameters matching the included LM somewhere, but why not as default values? 
At least some of the lesser changed values like `-arpa_prune` or `--binary_type`. 

It would keep the command with specific changes much more readable:
`python3 data/lm/generate_lm.py --input_txt path/to/vocab_sentences.txt --output_dir path/lm/ --arpa_order 3`
Instead of:
`python3 data/lm/generate_lm.py --input_txt path/to/vocab_sentences.txt --output_dir path/lm/ -top_k 500000 --kenlm_bins /DeepSpeech/native_client/kenlm/build/bin/ --arpa_order 3 --max_arpa_memory "75%" --arpa_prune "0|0|1" --binary_a_bits 255 --binary_q_bits 8 --binary_type trie`
In the above command you can see on the first look that i only changed the arpa order.

Can we ask @lissyx for opinion?

reuben replied:
For example, if you call the script with with `--arpa_order 2` you also have to modify the default pruning parameter. If we include defaults, people will run this script against their own data, modify some parameter, then the other defaults will conflict, and then they'll file an issue on our repo.

Besides, there's also the fact that all the default values included here are arbitrary, and only make sense for the default LM. There is no reason these values should be a default for any text corpus, so we should not misguide people using these scripts.

DanBmh replied:
Thanks for your explanation.

reuben commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,147 @@
-    with open('librispeech-vocab-500k.txt', 'w') as fout:
-      fout.write(vocab_str)
-
     # Filter LM using vocabulary of top 500k words
```
This comment is still out of date.

reuben commented on data/lm/generate_package.py:
```diff
@@ -128,7 +128,7 @@ def main():
         "--force_utf8",
         default="",
-        help="Boolean flag, force set or unset UTF-8 mode in the scorer package. If not set, infers from the vocabulary.",
+        help="Boolean flag, force set or unset UTF-8 mode in the scorer package. If not set, infers from the vocabulary. See <https://github.com/mozilla/DeepSpeech/blob/master/doc/Decoder.rst#utf-8-mode> for further explanation",
```
Thanks!

reuben commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,203 @@
+    lm_path = os.path.join(args.output_dir, "lm.arpa")
+    subprocess.check_call(
+        [
+            args.kenlm_bins + "lmplz",
```
`os.path.join`, not `+`

reuben commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,203 @@
+    filtered_path = os.path.join(args.output_dir, "lm_filtered.arpa")
+    subprocess.run(
+        [
+            args.kenlm_bins + "filter",
```
`os.path.join`, not `+`

reuben commented on data/lm/generate_lm.py:
```diff
@@ -1,63 +1,203 @@
+    binary_path = os.path.join(args.output_dir, "lm.binary")
+    subprocess.check_call(
+        [
+            args.kenlm_bins + "build_binary",
```
`os.path.join`, not `+`

DanBmh commented:
Could you please test the recent changes for me? I currently can't build a new deepspeech container.

DanBmh commented:
Tested now.

reuben commented on data/lm/README.rst:
```diff
@@ -1,8 +1,28 @@
-    python generate_lm.py # this will create lm.binary and librispeech-vocab-500k.txt
-    python generate_package.py --alphabet ../alphabet.txt --lm lm.binary --vocab librispeech-vocab-500k.txt --default_alpha 0.75 --default_beta 1.85 --package kenlm.scorer
+
+    python3 generate_package.py --alphabet ../alphabet.txt --lm lm.binary --vocab vocab-500000.txt \
+      --package kenlm.scorer --default_alpha 0.75 --default_beta 1.85
```
This needs to be updated with the rest of the values.
